### PR TITLE
Fix slow pyccel behaviour

### LINKED ISSUE #1114: Very slow processing of file with many functions
**Describe the bug**
When processing files that contain many functions, Pyccel has become very slow. Here is an example:

timings

**Expected behavior**
Pyccel version 1.5.0 should not be slower than version 0.10.1

**Language**
Fortran (tested), C (most likely)

### PR DESCRIPTION
**Pull request by EmilyBourne**
When calling `get_attribute_nodes` or `substitute` we must always exclude `FunctionDef` unless we are searching for a `FunctionDef`. Otherwise pyccel will search through all the contents of any functions called within the context. This is the cause of the bug in issue #1114 .

This PR fixes #1114 by redefining the class method `ignore` for `FunctionClass` so it never recurses into a `FunctionDef`.
Redundant ` excluded_nodes=(FunctionDef,)` commands are removed

### CODE REVIEW

File: pyccel/ast/basic.py [hunk 1]
```diff
@@ -63,12 +63,12 @@ def __init__(self):
...
+                    if not Basic._ignore(self, ci):
                         ci.set_current_user_node(self)
             else:
                 c.set_current_user_node(self)
 
-    def ignore(self, c):
+    def _ignore(self, c):
         """ Indicates if a node should be ignored when recursing
         """
         return c is None or isinstance(c, self._ignored_types)
```
Comment (yguclu) [inline pyccel/ast/basic.py:74]: Given the fact that `_ignored_types` is a class attribute, the behavior of this function does not depend on the `self` object. Hence I think it should be a class method (decorated with `@classmethod`).

File: pyccel/ast/core.py [hunk 2]
```diff
@@ -2050,6 +2050,11 @@ def interface_name(self):
     def __repr__(self):
         return '{}({})'.format(self.func_name, ', '.join(str(a) for a in self.args))
 
+    def _ignore(self, c):
+        """ Indicates if a node should be ignored when recursing
+        """
+        return c is None or isinstance(c, (FunctionDef, *self._ignored_types))
```
Comment (yguclu) [inline pyccel/ast/core.py:2056]: See previous comment: I think this should be a class method.

**Review by yguclu — CHANGES_REQUESTED** [review]
Great job!
I have only one comment

File: pyccel/ast/basic.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/ast/basic.py]: Fixed [6de7d35](https://github.com/pyccel/pyccel/pull/1113/commits/6de7d35e5ab13d6b0e04bd64cd7f0507cf511d9d)

File: pyccel/ast/basic.py [hunk 3]
```diff
@@ -51,7 +51,7 @@ def __init__(self):
             elif iterable(c):
                 size = len(c)
                 c = tuple(ci if (not isinstance(ci, (int, float, complex, str, bool)) \
-                                 or self._ignore(ci)) \
+                                 or Basic._ignore(ci)) \
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/basic.py:54]: In the `__init__` function it should be `Basic._ignore`. This ensures that the `user_node` is correctly set for all objects in `_attribute_nodes`.

In the other functions in the `Basic` class `self._ignore` should be used to stop the methods `substitute`, `invalidate_node`, `get_attribute_nodes`, `is_user_of` recursing into objects which should be ignored (i.e. `FunctionDef`)

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/basic.py:54]: Timings are the same (within very large error):
Run1:
```
real	0m17,087s
user	0m12,685s
sys	0m1,289s

real	0m12,016s
user	0m10,134s
sys	0m0,813s

real	0m19,586s
user	0m18,442s
sys	0m0,528s

real	0m55,924s
user	0m43,787s
sys	0m1,350s
```
Run2:
```
real	0m10,415s
user	0m9,711s
sys	0m0,684s

real	0m7,322s
user	0m7,022s
sys	0m0,333s

real	0m14,089s
user	0m13,694s
sys	0m0,390s

real	0m33,727s
user	0m33,287s
sys	0m0,427s
```

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/basic.py:54]: Run3 is similar to Run2 so this probably represents the timings when I am not running anything else on my computer

**Review by yguclu — APPROVED** [review]

**Comment by yguclu** [comment]
With Pyccel version 1.5.1. the processing time is 8.3 s, that is 67 % faster than version 0.10.1 which required 13.9 s.

**Pull request merged**